### DISCUSSION TITLE
Why Custom API rules make API slow ?

### DISCUSSION THREAD

**Discussion by samimhamza-pgl** [pocketbase/pocketbase#2400] [Q&A] [answered]
I have implemented permissions for restricting Pocketbase API's and when I add this rule for API, it make the API slow and it take much longer time.

>  @request.auth.id != "" && 
> @request.auth.permission_ids.name ?= "view vehicles" 

with this it takes around 4 seconds response time but without the permission it takes around 100 to 300 ms. 
I have also set name and permission_ids  as index but it doesn't change.

**Comment by samimhamza-pgl** [comment]
How would I be able to make the response time faster ?

**Reply by ganigeorgiev** [reply]
It is difficult to say what could be the reason for the slow query execution without any information about your dataset.

To answer your question why API rules may affect the performance - they are converted to SQL statements and depending on what actually you are checking the performance will vary.

- How large is your dataset (aka. how many records do you have in the search collection and the `permission_ids` collection)?

- Could you upload somewhere a sample `pb_data`?

- What is the version of PocketBase you are using?

- Do you have custom hooks that could interfere with the API response?

- Do you have `expand` with the query?

---

The easiest way to squeeze the performance is to use the CGO driver, aka. build PocketBase with `CGO_ENABLED=1 go build`.

Another thing that you can try is to add indexes.

To find the slow query, you can start PocketBase with `--debug=1` and check the `select` query that will be printed in the console when performing the list/search action.

Then you can use the sqlite3 cli tool and execute `EXPLAIN QUERY PLAN YOUR_QUERY` and see what the query planner is doing.

**Reply by samimhamza-pgl** [reply]
thanks for your response,

- My dataset is not that much big, and permission_ids is a column in users table with multiple relation on permissions table.
- permissions table has only four data for now and it may have 100 records till the end.
- The collection that I have applied this API rules is vehicles collection and it has around 80 thousands of records. but I only get 10 per page. and I have not used any expand till now.
- I am using v0.15.1 and I used it as a framework for implementing custom routes but I have not added any custom route till now.
- I am not using any custom hooks till now.
- I have added permission_ids from users collection and name from permissions table as index.

I'm not sure why there is too much difference in response time when I add any filter other than @request.auth.id != "" in API rules

**Reply by ganigeorgiev** [reply]
> My dataset is not that much big, and permission_ids is a column in users table with multiple relation on permissions table.

But how much big exactly - 1k, 100k, etc? Could you check the size of the db file in `pb_data/data.db`?

Could you paste the generated SELECT query?

Could you upload a minimal reproducible repo somewhere?

**Reply by samimhamza-pgl** [reply]
pb_data/data.db is about 147mb. 

the API that is sent to pocketbase from js-sdk is like this

> http://127.0.0.1:8090/api/collections/vehicles/records?page=1&perPage=10&sort=-created_at

I have tested this API rules with small collections also and it does not take too much time. 
but when I test it with vehicles collection that it has around 80 thousands records it take too much time.

`@request.auth.id != "" &&
@request.auth.role = "test" `

this is another test that I have implement but I get the same result.
role is a column in users collections and it take too much time when I add this to vehicles API rules but it take much less time with other collection.
I mean that this filter that I have added is not related to collection itself, it only check users collection. then why it take too much time on vehicles collections (large dateset) but less time on other collections (small dataset)

**Reply by ganigeorgiev** [reply]
1. If you can't provide a minimal reproducible repo, could you at least export your collections configuration as JSON (_Admin UI > Settings > Export collections_), so that I can try to fill a local db with dummy data?

2. If you are already using PocketBase as framework, could you start with `CGO_ENABLED=1 go run main.go serve` and see if there would be any difference?

**Reply by samimhamza-pgl** [reply]
https://drive.google.com/file/d/1-T3WFaXoVXq5zP1ivyitEBvEn8lcg6bM/view?usp=sharing

I have exported structure of project and you can download it from above link

**Reply by ganigeorgiev** [reply]
Thanks.

At first glance exploring the json, your project is fairly large (it has almost 50 fields in just the `vehicles` collection!), so I assume this may be one of the reason for the slow `@request.auth.*` checks because for any custom auth record schema field we perform a join to the main table (in your case `vehicles`) with the auth collection table in order to allow further dot-nation filtering.

I'll try to prepare some dummy data and will report later with my findings.

**Reply by samimhamza-pgl** [reply]
I appreciate your support, I'll wait for your response, thank you so much.

**Comment by ganigeorgiev** [comment, marked as answer]
**TL;DR**
Add index on the `created_at` column and the performance will be improved ~5x (from ~2.6s to ~450ms).
If you also build with `CGO_ENABLED=1` the performance improvements combined would be ~12x (from ~2.6s to ~210ms).

**Note that which index the query planner will decide to use will differ depending on what is the final request (whether there are additional client-side field filters, other sort attributes, etc.).**

---

My test setup:
- 10     `permissions` records named from "a"-"j"
- 1 000   `users` records with all 14 custom fields populated for each record (each user record has 3 permission relations)
- 100 000 `vehicles` records with all 46 custom fields populated for each record

The request I tested was:
http://127.0.0.1:8090/api/collections/vehicles/records?page=1&perPage=10&sort=-created_at


I performed several tests grouped in **2 categories** - with empty rule and rule with an auth relation filter.
For each the reported time is the average response time of 5 consequent requests.

To understand what is happening under the hood, I've included also the generated SQL queries.

### First without extra indexes.

**Empty filter:**
  - Filter: ""
  - Query:
    ```sql
    SELECT `vehicles`.*
    FROM `vehicles`
    ORDER BY `vehicles`.`created_at` DESC
    LIMIT 10
    ```
  - Reponse times:
    ```
    CGO_ENABLED=0 (default) ~200ms
    CGO_ENABLED=1           ~144ms
    ```


**Auth multiple relation filter:**

  - Filter:
    ```js
    @request.auth.id != "" &&
    @request.auth.permission_ids.name ?= "a"
    ```
  - Query:
    ```sql
    SELECT DISTINCT `vehicles`.*
    FROM `vehicles`
    LEFT JOIN `users` `__auth_users` ON `__auth_users`.`id`='evcp4x2ijq6j6h7'
    LEFT JOIN json_each(CASE WHEN json_valid(`__auth_users`.`permission_ids`) THEN `__auth_users`.`permission_ids` ELSE json_array(`__auth_users`.`permission_ids`) END) `__auth_users_permission_ids_je`
    LEFT JOIN `permissions` `__auth_users_permission_ids` ON `__auth_users_permission_ids`.`id` = `__auth_users_permission_ids_je`.`value`
    WHERE ('evcp4x2ijq6j6h7' != '' AND `__auth_users_permission_ids`.`name` = 'a')
    ORDER BY `vehicles`.`created_at` DESC
    LIMIT 10
    ```
  - Reponse times:
    ```
    CGO_ENABLED=0 (default) ~2570ms
    CGO_ENABLED=1           ~1240ms
    ```


### The same tests above **but with index on the `created_at` field**

**Empty filter:**
  ```
  CGO_ENABLED=0 (default) ~92ms
  CGO_ENABLED=1           ~78ms
  ```

**Auth releation filter:**
  ```
  CGO_ENABLED=0 (default) ~450ms
  CGO_ENABLED=1           ~210ms
  ```


---

### Questions you may ask:

**1. Why it is slower with the `@request.auth.permission_ids.name ?= "a"`?**
  Because it performs a JOIN with each `permission_ids` for each record from the `vehicles` collection, aka. internally it will do something like:

    
    Vehicle 1 - Permission a
    Vehicle 1 - Permission b
    ...
    Vehicle 1 - Permission j
     
    Vehicle 2 - Permission a
    Vehicle 2 - Permission b
    ...
    Vehicle 2 - Permission j
    
    and so on...
    

  This usually is not and issue with small tables, but in your case the `vehicles` table has 49 columns (46 schema fields + 3 system) and it could need some time to process all the data (especially the pure Go SQLite port; additionally there is a single-row join with the actual auth record that also has 14+ columns).

**2. How to know which index to add to improve the query performance?**

  This is difficult to anwer because it depends on the query/request you are executing.
  If you have the time, you can check the official SQLite [Query Planning](https://www.sqlite.org/queryplanner.html) and [Query Optimizer](https://www.sqlite.org/optoverview.html) overviews.

  Personally, I rely mostly on the output of the `EXPLAIN [[QUERY]]` to decide when to add an index.

  Alternatively, the [`sqlite3`](https://www.sqlite.org/cli.html) cli tool has a helper `.expert` mode command that would suggest what indexes you can add for your queries. So for example, when you are not sure what index to add, you can:

  ```sh
  # open the db with sqlite3
  $ sqlite3 ./pb_data/data.db

  sqlite> .expert
  sqlite> [PASTE YOUR QUERY]
  ```
  And you'll get an output like (the first statement is the index suggestion):
  ```sh
  CREATE INDEX vehicles_idx_1808786b ON vehicles(created_at DESC);

  SEARCH __auth_users_permission_ids USING INDEX idx_Vmr7e72 (name=?)
  SCAN vehicles USING INDEX vehicles_idx_1808786b
  SEARCH __auth_users USING INDEX sqlite_autoindex_users_1 (id=?) LEFT-JOIN
  SCAN __auth_users_permission_ids_je VIRTUAL TABLE INDEX 1: LEFT-JOIN
  USE TEMP B-TREE FOR DISTINCT
  ```
  
**3. Is there something that could be improved in PocketBase?**

There is definitely room for improvements. We can try to isolate the `@request.auth.*` only filters and execute them separately from the main query, but this will add fair amount of complexity and will require significant changes to the internals and at least for now it is a low priority.
I'll try after the ongoing work on the backup/restore apis and the goja app hooks to finally create a testbed with some benchmarks so that we have something to compare to and once this is done we could start tracking the performance with each release/feature and eventually think on fine-tuning the problematic filters/queries.

**Reply by ganigeorgiev** [reply]
_One minor side-note: the `invoices` collection list and view rules are invalid (it seems that the `@collection.*` collection that is used was probably deleted/renamed)_

**Reply by samimhamza-pgl** [reply]
Thank you so much for your time. It helped a lot.  Your support means a lot to us. Now it take much less response time only by adding created_at as index, before it was around 4 seconds but now it is around 200ms.
